internal/impl/kafka/franz_client.go

```diff
@@ -151,3 +156,92 @@ func FranzConnectionOptsFromConfig(conf *service.ParsedConfig, log *service.Logg
 	}
 	return d.FranzOpts(), nil
 }
+
+type kgoConnectionChecker struct {
+	brokerConnected  bool
+	unrecoverableErr error
+	mut              sync.Mutex
+}
+
+func (k *kgoConnectionChecker) OnBrokerConnect(_ kgo.BrokerMetadata, _ time.Duration, _ net.Conn, err error) {
+	k.mut.Lock()
+	defer k.mut.Unlock()
+
+	if err == nil {
+		k.brokerConnected = true
+		return
+	}
+
+	var ke *kerr.Error
+	if !errors.As(err, &ke) {
+		return
+	}
+
+	if !ke.Retriable {
+		k.unrecoverableErr = err
```

Comment:
You may want to use `kgo.IsRetryableBrokerErr`, which returns true for some other common problems (notably, temporary connection errors) and false for things that are truly not retryable.